lib/internal/Magento/Framework/Escaper.php

```diff
@@ -291,7 +291,7 @@ public function escapeUrl($string)
      */
     public function encodeUrlParam($string)
     {
-        return $this->getEscaper()->escapeUrl($string);
+        return $this->getEscaper()->escapeUrl((string)$string);
```

Comment:
I would recommend throwing a deprecation when the passed value is not a string. Casting anything to a string is a workaround and not a fix.

```
if (!is_string($string)) {
  trigger_error(sprintf('Argument #1 passed to %s:%s needs to be an string, but got %s. In future versions this will throw an error', __CLASS__, __METHOD__, gettype($string)), E_USER_DEPRECATED);
}

Reply:
Why do we add string casting for a method which was never intended to support anything but strings? https://github.com/magento/magento2/blob/94e140fc16ba280f80580ec8eba02e38cd11d98a/lib/internal/Magento/Framework/Escaper.php#L294 Adding proper type to method signature + fix incorrect usages seems more robust to me.

Reply:
Yes that's the wish of all developers. You need a little spot where this change has to be made known to the users. That's the idea behind the deprecation.

- Deprecate it
- On other version forbid that with typethint

Reply:
I totally agree with the workaround definition, until laminas 2.8.0 released.

Laminas 2.8.0 escaper is strict but the current callchain is not strict 100%, that's why our integer is auto coerced. So string conversion happens automatically instead of explicit way.

At the end we need a string because rawurlencode() works with strings. I guess we need to handle every argument pass as strict as we can, and @hostep 's suggestion is right that in RouteParamsResolver we can explicit cast url parts to strings.

Now i add one commit to just reverse the fast workaround, and somebody should add hints, or close the pr.  what should be the next step. Any help would be appreciated.

Reply:
as we can't declare argument types, it's a good idea to add deprecation notice. Unfortunately, it doesn't work for some reason. 

I'll continue working on it in https://github.com/magento/magento2/pull/33637